drivers/i2c/gpio_i2c_switch.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023 Ayush Singh <ayushdevel1325@gmail.com>
+ * Copyright (c) 2021 Jason Kridner, BeagleBoard.org Foundation
+ * Copyright (c) 2020 Innoseis BV
+ *
+ * Based on i2c_tca9656a.c
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gpio_i2c_switch
+
+#define GPIO_I2C_TRANSFER_DELAY 100
+
+#include <zephyr/kernel.h>
+#include <zephyr/device.h>
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/logging/log.h>
+#include <stdint.h>
+
+LOG_MODULE_REGISTER(gpio_i2c_switch, CONFIG_I2C_LOG_LEVEL);
+
+struct gpio_i2c_switch_config {
+	const struct device *bus;
+	const struct gpio_dt_spec gpio;
+};
+
+struct gpio_i2c_switch_data {
+	struct k_mutex lock;
+};
+
+static int gpio_i2c_switch_configure(const struct device *dev, uint32_t dev_config)
+{
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	return i2c_configure(config->bus, dev_config);
+}
+
+static int gpio_i2c_switch_enable(const struct device *dev)
+{
+	int res;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_HIGH);
+
+	return res;
+}
+
+static int gpio_i2c_switch_disable(const struct device *dev)
+{
+	int res;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = gpio_pin_configure(config->gpio.port, config->gpio.pin, GPIO_OUTPUT_LOW);
+
+	return res;
+}
+
+static int gpio_i2c_switch_transfer(const struct device *dev, struct i2c_msg *msgs,
+				    uint8_t num_msgs, uint16_t addr)
+{
+	int res;
+	struct gpio_i2c_switch_data *data = dev->data;
+	const struct gpio_i2c_switch_config *config = dev->config;
+
+	res = k_mutex_lock(&data->lock, K_MSEC(5000));
+	if (res) {
+		return res;
+	}
+
+	gpio_i2c_switch_enable(dev);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY);
+	res = i2c_transfer(config->bus, msgs, num_msgs, addr);
+	k_busy_wait(GPIO_I2C_TRANSFER_DELAY);
+	gpio_i2c_switch_disable(dev);
+
+	k_mutex_unlock(&data->lock);
+	return res;
```

Comment:
Either add a label and goto from the error condition above, or change this to `return 0;`, in any case handle the return value close to where it's set, that is check for it after `i2c_transfer`